File: 1-2-procedures-and-processes-they-generate/1-2-1-linear-recursion-and-iteration/factorial.hs
-- SICP 1.2.1
-- Linear Recursion and Iteration

fact1 :: Int -> Int
fact1 1 = 1
fact1 n = n * fact1(n - 1)


-- Iterative

fact2 :: Int -> Int
fact2 n = fact2iter 1 1 n

fact2iter :: Int -> Int -> Int -> Int
fact2iter product counter maxcount
| counter > maxcount = product
| otherwise = fact2iter (counter*product) (counter+1) maxcount


main = do
  print (fact1 7)
  print (fact2 7)

File: 1-2-procedures-and-processes-they-generate/1-2-1-linear-recursion-and-iteration/factorial.swift
// SICP 1.2.1
// Linear Recursion and Iteration

func fact1(for n: Int) -> Int {
    guard n != 1 else { return 1 }
    return n * (fact1(for: (n-1)))
}

// Iterative

func fact2(for n: Int) -> Int {
    func fact2Iter(_ product: Int, _ counter: Int, _ maxCount: Int) -> Int {
        if counter > maxCount {
            return product
        } else {
            return fact2Iter(counter*product, counter+1, maxCount)
        }
    }
    
    return fact2Iter(1, 1, n)
}

print(fact1(for: 7))
print(fact2(for: 7))